libethereum/State.cpp

```diff
@@ -535,57 +535,54 @@ pair<TransactionReceipts, bool> State::sync(BlockChain const& _bc, TransactionQu
 					if (req > got)
 					{
 						// too old
-						for (Transaction const& t: m_transactions)
-							if (t.from() == i.second.from())
+						for (Transaction const& mt: m_transactions)
+						{
+							if (mt.from() == t.from())
 							{
-								if (t.nonce() < i.second.nonce())
-								{
-									cnote << i.first << "Dropping old transaction (nonce too low)";
-									_tq.drop(i.first);
-								}
-								else if (t.nonce() == i.second.nonce() && t.gasPrice() <= i.second.gasPrice())
-								{
-									cnote << i.first << "Dropping old transaction (gas price lower)";
-									_tq.drop(i.first);
-								}
+								if (mt.nonce() < t.nonce())
+									cnote << t.sha3() << "Dropping old transaction (nonce too low)";
+								else if (mt.nonce() == t.nonce() && mt.gasPrice() <= t.gasPrice())
+									cnote << t.sha3() << "Dropping old transaction (gas price lower)";
 							}
+						}
+						_tq.drop(t.sha3());
```

Comment:
There are quite a lot of transactions coming from the network with outdated nonces (less than last block account state).  This function gets transactions ordered by nonce and tries to import them. If  an exception is thrown with nonce being too old, this means it won't succeed even if the state is cleared. It can only succeed if a blockchain is reorganized to a previous block, but his case is handled in Client.cpp